parlai/tasks/coco_caption/agents.py

```diff
@@ -0,0 +1,301 @@
+        annotation_suffix = 'val{}'.format(version)
+        img_suffix = os.path.join('val{}'.format(version),
+                                  'COCO_val{}_'.format(version))
+    elif dt == 'test':
+        annotation_suffix = 'None'
+        img_suffix = os.path.join('test{}'.format(version),
+                                  'COCO_test{}_'.format(version))
+    else:
+        raise RuntimeError('Not valid datatype.')
+
+    test_info_path = os.path.join(opt['datapath'],
+                                  'COCO_{}_Caption'.format(version),
+                                  'annotations',
+                                  'image_info_test{}.json'.format(version))
+
+    annotation_path = os.path.join(opt['datapath'],
+                                   'COCO_{}_Caption'.format(version),
+                                   'annotations',
+                                   'captions_' + annotation_suffix + '.json')
+
+    image_path = os.path.join(opt['datapath'], 'COCO-IMG', img_suffix)
+
+    return test_info_path, annotation_path, image_path
+
+
+class COCODataset(Dataset):
+    """A Pytorch Dataset utilizing streaming"""
+    def __init__(self, opt):
+        self.opt = opt
+        self.use_att = opt.get('attention', False)
+        self.use_hdf5 = not opt.get('no_hdf5', False)
+        self.datatype = self.opt.get('datatype')
+        self.training = self.datatype.startswith('train')
+        self.num_epochs = self.opt.get('num_epochs', 0)
+        self.image_loader = ImageLoader(opt)
+        test_info_path, annotation_path, self.image_path = _path(opt)
+        self._setup_data(test_info_path, annotation_path, opt.get('unittest', False))
+        if self.use_hdf5:
+            try:
+                import h5py
+                self.h5py = h5py
+            except ModuleNotFoundError:
+                raise ModuleNotFoundError('Need to install h5py - `pip install h5py`')
+            self._setup_image_data()
+        self.dict_agent = VqaDictionaryAgent(opt)
+
+    def __getitem__(self, index):
+        index %= self.num_episodes()
+        image_id = None
+        if not self.datatype.startswith('test'):
+            anno = self.annotation['annotations'][index]
+            image_id = anno['image_id']
+        else:
+            image_id = self.test_info['images'][index]['id']
+        ep = {
+            'text': QUESTION,
+            'image': self.get_image(image_id),
+            'episode_done': True,
+        }
+        if self.opt.get('extract_image', False):
+            ep['image_id'] = image_id
+            return ep
+        if not self.datatype.startswith('test'):
+            anno = self.annotation['annotations'][index]
+            ep['labels'] = [anno['caption']]
+            ep['valid'] = True
+            if 'mc_label' in ep:
+                if not ep['mc_label'][0] in self.dict_agent.ans2ind:
+                    ep['valid'] = False
+            ep = self.dict_agent.encode_question([ep], self.training)
```

Comment:
with the new dict agent, these should be changed to match the functions in `DictAgent`